File: core/project.py
"""Project detection: CWD → git root → bank name.

Bank naming:
- Inside a git repo → bank is the sanitised repo directory name (e.g. 'credo_main')
- Inside $HOME git repo → falls through to 'system' (dotfiles repo)
- Outside a git repo → 'system'

Resolution order for working directory:
1. HINDSIGHT_CWD env var (set by MCP config or agent hooks)
2. cwd argument (if provided)
3. os.getcwd()
"""

from __future__ import annotations

import os
from pathlib import Path

SHARED_BANK = "system"


def resolve_working_dir() -> Path:
    """Get the current working directory.

    Checks HINDSIGHT_CWD env var first — this lets MCP server configs
    pass the agent's project directory even when the server process
    starts in a different directory (e.g. the install/lib dir).
    """
    env_cwd = os.environ.get("HINDSIGHT_CWD", "").strip()
    if env_cwd:
        p = Path(env_cwd).resolve()
        if p.is_dir():
            return p
    return Path(os.getcwd()).resolve()


def detect_project(cwd: Path | None = None) -> str:
    """Derive a project name from CWD by walking up to find .git.

    Returns the git repo directory name if inside a git repo.
    Returns 'system' for anything outside a git repo (home dir, /tmp, etc.).
    """
    if cwd is None:
        cwd = resolve_working_dir()

    resolved = cwd.resolve()
    home = _home()

    for parent in [resolved, *resolved.parents]:
        if (parent / ".git").exists():
            # Skip if git root is $HOME (dotfiles repo) — treat as system
            if home and parent == home:
                continue
            return sanitise(parent.name)

    return SHARED_BANK


def sanitise(value: str) -> str:
    """Sanitise a string for use as a bank ID.

    Converts special characters to hyphens, collapses runs,
    strips leading/trailing hyphens, and lowercases.
    """
    if not value:
        return ""
    out = []
    prev_dash = False
    for ch in value:
        if ch.isalnum() or ch in "-_":
            out.append(ch)
            prev_dash = False
        elif not prev_dash:
            out.append("-")
            prev_dash = True
    return "".join(out).strip("-_").lower()


def _home() -> Path | None:
    """Return resolved $HOME, or None if unavailable."""
    try:
        return Path.home().resolve()
    except (OSError, RuntimeError):
        return None

File: integrations/opencode/src/project.ts
/**
 * Project detection: CWD → git root → bank name.
 *
 * Same logic as core/project.py — ensures conformity across all agents.
 * - Inside a git repo → bank is the sanitised repo directory name
 * - Inside $HOME git repo → falls through to 'system'
 * - Outside a git repo → 'system'
 */

import { resolve, basename, dirname, join } from "node:path";
import { existsSync } from "node:fs";
import { execFileSync } from "node:child_process";
import { homedir } from "node:os";

export const SHARED_BANK = "system";

/**
 * Derive a project name from a directory by walking up to find .git.
 */
export function detectProject(directory: string): string {
    const resolved = resolve(directory);
    const home = resolve(homedir());

    let current = resolved;
    while (true) {
        if (existsSync(join(current, ".git"))) {
            // Skip if git root is $HOME (dotfiles repo)
            if (current === home) {
                const parent = dirname(current);
                if (parent === current) break;
                current = parent;
                continue;
            }
            return sanitise(basename(current));
        }
        const parent = dirname(current);
        if (parent === current) break;
        current = parent;
    }

    return SHARED_BANK;
}

/**
 * Resolve the main worktree root for a directory inside a git repository.
 * All linked worktrees of the same repo share one memory bank.
 */
export function getGitProjectRoot(directory: string): string | null {
    if (!directory) return null;
    try {
        const commonDir = execFileSync(
            "git",
            ["rev-parse", "--path-format=absolute", "--git-common-dir"],
            { cwd: directory, encoding: "utf-8", stdio: ["ignore", "pipe", "ignore"], timeout: 1000 }
        ).trim();
        if (!commonDir) return null;
        return dirname(commonDir);
    } catch {
        return null;
    }
}

/**
 * Sanitise a string for use as a bank ID.
 */
export function sanitise(value: string): string {
    if (!value) return "";
    let out = "";
    let prevDash = false;
    for (const ch of value) {
        if (/[a-zA-Z0-9\-_]/.test(ch)) {
            out += ch;
            prevDash = false;
        } else if (!prevDash) {
            out += "-";
            prevDash = true;
        }
    }
    return out.replace(/^[-_]+|[-_]+$/g, "").toLowerCase();
}
